src/components/input/InputRoot.tsx

```diff
@@ -1,43 +1,48 @@
 import classnames from 'classnames';
+import type { JSX } from 'preact';
 
+import type { PresentationalProps } from '../../types';
 import { downcastRef } from '../../util/typing';
 
 import { inputGroupStyles } from './InputGroup';
 
-/**
- * @typedef {import('../../types').PresentationalProps} CommonProps
- * @typedef {import('preact').JSX.HTMLAttributes<HTMLInputElement>} HTMLInputAttributes
- *
- * @typedef InputProps
- * @prop {boolean} [hasError=false]
- * @prop {'email'|'search'|'text'|'url'} [type="text"]
- */
+type RootComponentProps = {
+  element?: 'input' | 'select';
+  hasError?: boolean;
+};
+
+export type InputRootProps = PresentationalProps &
+  RootComponentProps &
+  (
+    | Omit<JSX.HTMLAttributes<HTMLInputElement>, 'size' | 'icon'>
+    | Omit<JSX.HTMLAttributes<HTMLSelectElement>, 'size' | 'icon'>
+  );
 
 /**
- * Render a text field input
- *
- * @param {CommonProps & InputProps & Omit<HTMLInputAttributes, 'type'>} props
+ * Root component for various input types that applies a consistent design
+ * pattern.
  */
-const InputNext = function Input({
+const InputRootNext = function InputRoot({
+  element: Element = 'input',
   children,
   classes,
   elementRef,
-
   hasError,
-  type = 'text',
 
   ...htmlAttributes
-}) {
+}: InputRootProps) {
   // @ts-expect-error - "aria-label" is missing from HTMLInputAttributes
   if (!htmlAttributes.id && !htmlAttributes['aria-label']) {
     console.warn(
       '`Input` component should have either an `id` or an `aria-label` attribute'
     );
   }
+
   return (
-    <input
+    <Element
+      data-component="InputRoot"
       {...htmlAttributes}
-      type={type}
+      // @ts-ignore-next
```

Comment:
Ref-typing purgatory...I should add more comment here but I don't really know what to say.